docs/proposal/content-library-support.md

```diff
@@ -0,0 +1,134 @@
+
+Using content library as the source of templates has another little benefit that enables using local datastore as destination for the VMs, currently we are limited to shared datastore.
+
+## Motivation
+
+Enhance CAPV to clone VM from content library, to make multi-vCenter template management easier.
+
+### Goals
+
+* Use ContentLibrary items as the VM image template.
+
+### Non-Goals
+
+* Turn content library the only supported template source.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an admin, I'd like to manage templates/OVAs across the vCenters that can be consumed by CAPV.
+
+#### Story 2
+
+As an cluster user, I'd like to use content library items(templates/OVAs) as the source of cluster nodes.
+
+### Overall Design
+
+The design focuses on how to clone VMs from content library items. For the content library management itself users can reference vSphere documentation.
+
+### Implementation Details/Notes/Constraints
+
+#### Naming scheme
+
+We use `Template` field of `VirtualMachineCloneSpec` to specify the source we want to clone,
+since we want to augment it to also from the content library, we need to have a way to differentiate the source.
+I propose we use URI scheme for that, it has the format: `scheme://location` which
+
+* scheme, can be vsphere|library
+  * vsphere for conventional vm templates
+  * library for OVA/templates from content library
+* to make it backward compatible, the scheme part can be omitted for conventional vm templates, so if we encounter
+  template without scheme part, we assume it is vsphere template, existing CAPV does not need to do anything.
+
+#### API changes
+
+##### VirtualMachineCloneSpec CR
+
+The `VirtualMachineCloneSpec` has following changes, the spec is used to specify clone spec for `VsphereVM`, it needs the following change to deal with newly added content library capabilities.
+
+```golang
+type VirtualMachineCloneSpec struct {
+  // Template is the name or inventory path of the template used to clone
+  // the virtual machine. It can be a vSphere template, or a content library item. For example:
+  // vsphere template: [vsphere://location of the vsphere template], vsphere:// can be omitted for backward compatibility
+  // content library: [library://name of the content library item]
+  // note that this has impact on cloneMode field, since it is only possible to do full clone from content library items,
+  // and we should have validation rule for this.
+  // +kubebuilder:validation:MinLength=1
+  Template string `json:"template"`
+
+  // The secret that stores the credentials for accessing the content library.
+  // It should reside on the same namespace of the Cluster that reference it.
+  // It is optional since enabling access control for the content library is optional. It's up to content library admin to decide if
+  // it is required to enable access control.
+  // +optional
+  LibraryCredentials string `json:"library_credentials"`
```

Comment:
Should we make this of type `corev1.TypedLocalObjectReference`?

This way we would be able to in future support other object types too without having to adjust the API.

Reply:
Why are credentials for the library needed? A sequence diagram of interactions between VA-API and CLS that go from library item to VM creation should make this clear.

The content library will already have been subscribed to in vCenter with whatever http credentials it needs, we should only need to pass it's UUID to CreateTemplate or CreateOVF

Reply:
re-reading the doc, as the overall design states that `For the content library management itself users can reference vSphere documentation`, I agree the credential should not be required.

Also, we should probably add that the "Management of content library items and subscriptions" is a nongoal, so it is clear that the VI admin should first subscribe/create a content library and CAPV will only use it.